compiler/plugins/target/AMD-AIE/aievec/VectorToVectorConversions.cpp

```diff
@@ -364,6 +369,222 @@ struct CanonicalizeTrivialReadAccessSubviewOpPattern
+    : public OpRewritePattern<vector::TransferWriteOp> {
+  using OpRewritePattern<vector::TransferWriteOp>::OpRewritePattern;
+  SerializeSplatTransferReadWithTargetLoadSize(MLIRContext *context,
+                                               AMDAIE::AMDAIEDevice d)
+      : OpRewritePattern<vector::TransferWriteOp>(context),
+        deviceModel(AMDAIE::getDeviceModel(d)) {}
+
+ private:
+  AMDAIE::AMDAIEDeviceModel deviceModel;
+  LogicalResult matchAndRewrite(vector::TransferWriteOp writeOp,
+                                PatternRewriter &rewriter) const override {
+    Value writeDestination = writeOp.getSource();
+    MemRefType writeDestinationType =
+        dyn_cast<MemRefType>(writeDestination.getType());
+
+    assert(writeDestinationType && "transfer_write must write to memref");
+    if (!writeDestinationType.hasStaticShape()) {
+      return rewriter.notifyMatchFailure(writeOp, "source has dynamic shape");
+    }
+    {
+      MemRefLayoutAttrInterface layout = writeDestinationType.getLayout();
+      if (layout && !layout.isIdentity()) {
+        return rewriter.notifyMatchFailure(writeOp,
+                                           "source has non-identity layout");
+      }
+    }
+    // This pass only succeeds when the memref that is written to is rank-1.
+    // There is already a pattern for flattening memrefs that takes care of
+    // rank-n (n>1) memrefs.
+    if (writeDestinationType.getRank() != 1) {
+      return rewriter.notifyMatchFailure(writeOp, "memref isn't rank-1");
+    }
+
+    Operation::operand_range initialWriteIndices = writeOp.getIndices();
+    assert(initialWriteIndices.size() == 1 && "write destination is rank-1");
+    Attribute initialIndexAttr;
+    matchPattern(initialWriteIndices[0], m_Constant(&initialIndexAttr));
+    if (!initialIndexAttr) {
+      return rewriter.notifyMatchFailure(writeOp, "index isn't constant");
+    }
+    int64_t initialIndex = cast<IntegerAttr>(initialIndexAttr).getInt();
+
+    VectorType initialVectorType =
+        dyn_cast<VectorType>(writeOp.getVector().getType());
+    // Sanity checks on the vector operand:
+    assert(initialVectorType && "vector must be of vector type");
+    assert(writeDestinationType.getElementType() ==
+               initialVectorType.getElementType() &&
+           "element types must match");
+
+    // Find the arith.constant that the vector operand is a view of, if it is
+    // one.
+    arith::ConstantOp constantVectorSource = [&writeOp]() -> arith::ConstantOp {
+      Value current = writeOp.getVector();
+      while (Operation *op = current.getDefiningOp()) {
+        if (auto cOp = dyn_cast<arith::ConstantOp>(op)) return cOp;
+        if (op->getNumOperands() != 1) return {};
+        current = op->getOperand(0);
+      }
+      return {};
+    }();
+    if (!constantVectorSource) {
+      return rewriter.notifyMatchFailure(
+          writeOp, "vector isn't derived from arith.constant");
+    }
```

Comment:
The [comment on `maybeSplat`](https://github.com/nod-ai/iree-amd-aie/pull/1095/files#r1964892999) applies here as well. No need to define such inlined functions when it's really just invoked once.

Reply:
I'll change it, but I actually find the approach with lambdas clearer to read: (1) my eye can skip to end of function if it's not interested in it's details of the traversal (2) name encapsulation.

Reply:
Ok, done! It does look nice now.